src/persist-txn/src/txn_read.rs

```diff
@@ -137,26 +143,106 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64, C: TxnsCodec>
     /// Callers must first wait for `update_gt` with the same or later timestamp
     /// to return. Panics otherwise.
     ///
-    /// Returns Err if that data has not been registered.
-    pub fn to_data_inclusive(&self, data_id: &ShardId, ts: T) -> Result<T, NotRegistered<T>> {
-        assert!(self.progress_exclusive > ts);
-        self.datas
-            .get(data_id)
-            .and_then(|all| {
-                let ret = all.iter().rev().find(|x| **x <= ts).cloned();
-                debug!(
-                    "to_data_inclusive {:.9} t={:?} ret={:?}: all={:?}",
-                    data_id.to_string(),
-                    ts,
-                    ret,
-                    all,
-                );
-                ret
-            })
-            .ok_or(NotRegistered {
-                data_id: *data_id,
-                ts: self.progress_exclusive.clone(),
-            })
+    /// Returns Err if that data shard has not been registered at the given ts.
+    pub fn data_snapshot(
+        &self,
+        data_id: &ShardId,
+        as_of: T,
+    ) -> Result<DataSnapshot<T>, NotRegistered<T>> {
+        assert!(self.progress_exclusive > as_of);
+        let all = self.datas.get(data_id).ok_or(NotRegistered {
+            data_id: *data_id,
+            ts: self.progress_exclusive.clone(),
+        })?;
+        let latest_write = all
+            .iter()
+            .rev()
+            .find(|x| **x <= as_of)
+            .unwrap_or(&as_of)
+            .clone();
+        let empty_to = all
```

Comment:
I probably need to talk this through in person, because the concept of `empty_to` breaks down for me when the `as_of` is "in between" a bunch of timestamps at which this shard has writes. Say we have writes at `1, 2, 3, 4, 5`. Now I request a shapshot at `as_of=3`. Then `latest_write=3` and `empty_to=4`. Now, though, `unblock_read()` doesn't have to do any writes, it's a no-op? Because the "emptyness" that's between `[latest_write, empty_to)` has stuff on either sides.

Whenever `empty_to` is an element of `all` and not actually `progress_exclusive`, we're in this situation. Is that correct?

I think this, and the interplay with `unblock_reads` is correct. But it took me a while to understand it. Probably because this is trying to cover more nuance that I'm missing?